### PR TITLE
Build with latest/stable charmcraft (requirements.txt without hashes)

### PR DESCRIPTION
**Pull request by paulomach**
## Issue

* charmcraft pinned to `2.3.0`
* charmcraft stable (`>2.3.0, <=2.4.1`) can't do pip's secure installs (`requirements.txt` with hashes).

## Solution

* unpin charmcraft
* poetry export `requirements.txt` w/o hashes

### CODE REVIEW

**Review by taurus-forever — DISMISSED** [review]
All tests passed, LGTM! No need to pin old charmcraft without a real reason :-)

File: tox.ini [hunk 1]
```diff
@@ -24,7 +24,7 @@ allowlist_externals =
     charmcraft
     mv
 commands_pre =
-    poetry export --only main,charm-libs --output requirements.txt
+    poetry export --only main,charm-libs --without-hashes --output requirements.txt
```
[the remark this answers is not in the document]
Comment (paulomach) [reply tox.ini]: dc92b62

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -34,9 +34,7 @@ jobs:
 
   build:
     name: Build charm
-    uses: canonical/data-platform-workflows/.github/workflows/build_charms_with_cache.yaml@v4.2.3
-    with:
-      charmcraft-snap-revision: 1349  # version 2.3.0
+    uses: canonical/data-platform-workflows/.github/workflows/build_charms_with_cache.yaml@v4.2.0
```
[the remark this answers is not in the document]
Comment (paulomach) [reply .github/workflows/ci.yaml]: :man_facepalming: no reason, bad conflict solving 

dc92b62

**Comment by paulomach** [comment]
> Is there any specific reason to use the latest charmcraft (i.e. feature or bug fix needed)? If not, maybe it'd be better to keep it pinned so that we don't lose the security benefit of hashes: https://pip.pypa.io/en/stable/topics/secure-installs/#hash-checking-mode

It's a tradeoff between security and maintenance effort. @taurus-forever it's your call

**Comment by taurus-forever** [comment]
> Is there any specific reason to use the latest charmcraft (i.e. feature or bug fix needed)?
> If not, maybe it'd be better to keep it pinned so that we don't lose the security benefit of hashes:...

@carlcsaposs-canonical you are absolutely right about security concerns here!
This is why charmcraft needs to react seriously to bugreport https://github.com/canonical/charmcraft/issues/1179 + https://github.com/canonical/charmcraft/issues/1140 (sent ping there).

The reason why I am asking to fix it this way was the last interview candidate, which reported:
"I have tried to compile your charm following CONTRIBUTING.md, but it failed with strange errors I was not able to understand. I gave up...". Investigation showed:  charmcraft=2.4.1 from latest/stable.

The pinning of old version is a hidden knowledge (visible in CI/CD only).
The charmcraft 2.4.1 is the latest/stable. All team-mates/community will use it to build our charm (just because `snapd` refresh it silently).

My position: temporary remove hashes, until charmcraft releases the fix.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -35,8 +35,6 @@ jobs:
   build:
     name: Build charm
     uses: canonical/data-platform-workflows/.github/workflows/build_charms_with_cache.yaml@v4.2.3
-    with:
-      charmcraft-snap-revision: 1349  # version 2.3.0
```
Comment (taurus-forever) [inline .github/workflows/ci.yaml:39]: [here](https://github.com/canonical/mysql-k8s-operator/blob/main/.github/workflows/release.yaml#L38) as well? (c) Carl

**Review by taurus-forever — DISMISSED** [review]

**Comment by paulomach** [comment]
no longer topical

**Pull request closed without merging**